### DISCUSSION TITLE
GET /offers/listing dla wszystkich, bez informacji "wrażliwych"

### DISCUSSION THREAD

**Discussion by juhacz** [allegro/allegro-api#4584] [Suggestions - rejected]
Dzień dobry
Zgłaszam sugestię dotyczącą stworzenia GET /offers/listing dla wszystkich.
Allegro zabrało użytkownikom ten zasób ponieważ uważa, że zwraca informacje, które w jakiś sposób uważa za wrażliwe.  Zasób udostępnia tylko wybranym. Wiele użytkowników potrzebuje zasobu, który pozwoli im pobierać nowo wystawione aukcje, a jedyny zasób, który to umożliwiał, to właśnie ten, REST Api nie ma dziennika jak WebAPI, więc nie ma innej możliwości. 
**Dlatego proponuję, aby GET /offers/listing niezweryfikowanym aplikacjom zwracał dane bez wrażliwych danych** (podejrzewam, że chodzi o popularność), a tym zweryfikowanym wszystkie. Wydaje mi się, że byłoby to uczciwe, ponieważ takim postępowaniem „ubiliście” wiele biznesów o czym było tu już słychać wiele razy.
Ta całą sytuacja przypomina mi pewną analogię, zakazujemy sprzedaży w noży w sklepach, ponieważ  ktoś może nimi zabić sąsiada, ale 99,999999999% innych klientów potrzebuje go w kuchni.

**Comment by fakk3** [comment]
Miejmy nadzieję, że allegro zwróci na to uwagę. W obecnej formie ręczne przeglądanie ofert to jakaś masakra.

**Reply by DarkStoorM** [reply]
Trochę dziwne, że listing został wyłączony dla "mniejszych", skoro publiczna wyszukiwarka jest dostępna dla każdego. Nie mogli po prostu dla "niezweryfikowanych" aplikacji zwrócić okrojonej wersji? To tak, jakby na froncie wyszukiwarka była dostępna dla kupujących, którzy wydają min 50 tysi miesięcznie. Scrapper i tak działa, więc nie wiem co to dało ¯\_(ツ)_/¯

**Comment by hahyes** [comment]
Szkoda, że traktują sugestie w sposób tak wybiórczy i nie chce się im nawet odpowiedzieć, że nie czy cokolwiek.

Ale nie łudziłbym się - te ograniczenia powstały po to, żeby mniejsi sprzedawcy utracili dostęp do wyszukiwarki przez API i korzystali z usług analitycznych Allegro za grubą kasę. Dużym pozostawili tą możliwość, ponieważ to są duże pieniądze, więc wiadomo. Reszcie bardziej się opłaca wyszukiwarkę zabrać i tyle.

A że scrapper działa... może lepiej się nie chwalcie?

**Reply by AureliuszBrussy** [reply]
W zakładce Discussions zbieramy Wasze potrzeby - czekamy także na Wasze głosy w zaproponowanych funkcjonalnościach, aby określić priorytet danej propozycji. Sugestie cyklicznie przekazujemy dalej, a jeżeli zdecydujemy się na wdrożenie danej propozycji - informujemy statusie prac w danym wątku. Informujemy także o ewentualnym odrzuceniu propozycji.

**Reply by hahyes** [reply]
> W zakładce Discussions zbieramy Wasze potrzeby - czekamy także na Wasze głosy w zaproponowanych funkcjonalnościach, aby określić priorytet danej propozycji. Sugestie cyklicznie przekazujemy dalej, a jeżeli zdecydujemy się na wdrożenie danej propozycji - informujemy statusie prac w danym wątku. Informujemy także o ewentualnym odrzuceniu propozycji.

Rozumiem. Myślałem, że wygląda to nieco inaczej. Zwracam honor.

**Comment by maciekjaros** [comment]
Popieram, dla niezweryfikowanych aplikacji można okroić odpowiedź o te "wrażliwe" dane.

**Comment by meridPL** [comment]
Jeżeli nie można udostępniać danych wrażliwych jednym, to drugim również. Dane wrażliwe to dane wrażliwe nie można jednym je udostępniać, a drugim nie. Kto ma pieniądze ma dane, kto nie ma niech spada. Nie tędy droga. W pewnym momencie i tak wrócicie do tego API, bo konkurencja nie śpi. Więcej takich decyzje spowoduje przejście do kogoś innego. Ja wiem, że tego prezes nie czyta tylko pionki ustawione przez niego, ale może ruszycie głową, bo jak nie będzie klientów, to też nie będzie pracy ;) Już tego nie unikniecie, bo ktoś z bardzo wielkim kapitałem chce to zmienić. I tu radzę sobie przypomnieć firmy Optimus lub Nasza Klasa.

Pozdrawiam wszystkich i mam nadzieję, że nie obudzicie się z mokrymi rękami ;)

**Reply by hahyes** [reply]
O jakiej konkurencji piszesz? Amazon z chińskim towarem, średniowiecznym UI i kosmicznymi prowizjami? eBay, które umarło w momencie pojawienia się na polskim rynku? Czy o hipotetycznej platformie Comarch, która równie dobrze może być kolejnym wyssaniem kasy z Unii i zrobią jakąś platformę w 5 minut, jak było z wszystko peel? Gdzie sprzedawcy przejdą niby? Do kogo? Allegro rządzi na rynku e-commerce i niestety mogą dzięki tej pozycji robić co chcą.

Konkurencja e-commerce w Polsce nie tylko śpi, ale regularnie stacza się z łóżka i uderza głową o podłogę.

**Reply by maciekjaros** [reply]
Powiedzmy sobie szczerze, że to nie ochrona danych tylko zwykłe praktyki monopolistyczne. Więc zamiast pisać śmieszne prośby o przywrócenie dostępu do API, które rozpatrzone zostaną w kompletnie nieokreślonej przyszłości (bądź nie), napiszę chyba do UOKiK, żeby zbadali sprawę nadużywania dominującej pozycji :)

**Reply by hahyes** [reply]
> Powiedzmy sobie szczerze, że to nie ochrona danych tylko zwykłe praktyki monopolistyczne. Więc zamiast pisać śmieszne prośby o przywrócenie dostępu do API, które rozpatrzone zostaną w kompletnie nieokreślonej przyszłości (bądź nie), napiszę chyba do UOKiK, żeby zbadali sprawę nadużywania dominującej pozycji :)

Nie przejdzie.
1) Zasób pozostał dla dużych i zweryfikowanych sprzedawców, ale jego możliwości zostały mocno ograniczone.
2) Allegro nie ma w obowiązku udostępniać wszystkich zasobów byle komu. Choćby "konkurencyjny" Amazon ma kilka poziomów dostępu do API i żeby wejść wyżej trzeba spełnić warunki, tak jak tutaj.
3) "Dane wrażliwe" to jest nadal dobry argument.

**Reply by maciekjaros** [reply]
Pewnie nie. Ale zgłosić można, nic nie tracę. Lecę teraz na scrapperze przez proxy, ale działa to zdecydowanie gorzej.

**Comment by adamz464** [comment]
Niech API będzie płatne rozliczane wg zużycia. Ktoś kto ma na własny użytek to go nie zaboli, a ktoś kto robi aplikacje dla innych to uwzględni w abonamencie.

**Reply by meridPL** [reply]
Najgłupszy pomysł jaki może być. Mało kasy trzepią ?

**Reply by juhacz** [reply]
> Niech API będzie płatne rozliczane wg zużycia. Ktoś kto ma na własny użytek to go nie zaboli, a ktoś kto robi aplikacje dla innych to uwzględni w abonamencie.

Już kiedyś w WebApi były niektóre metody płatne, zrezygnowali z tego, więc to im kasy nie przynosi.

**Reply by adamz464** [reply]
skoro zabrali to wierzysz, że za darmo przywrócą?
Jedyny argument to kasa, a nie żale. Sobie pisać można i tak to nic nie da.

Łasi na kasę są strasznie - np ostatnie zmiany 15% w kategoriach pozostałych dla firm, a dla prywatnych 9%. Jak by im nie zależało to by nie podnosili.

Taki ebay chociaż daje RSS bez żadnej łaski. Może prędzej taka idea aniżeli ta jedna metoda API.

**Comment by adamz464** [comment]
Napiszę osobno tutaj. Zamiast walczyć o tę metodę to lepiej o mechanizm RSS jak na Ebay. Dla powiadomień o aukcjach, jest to w zupełności wystarczające i odpadają wszelkie argumenty o danych wrażliwych.

**Reply by maciekjaros** [reply]
Ale jak chcesz odpowiednio często sprawdzać RSS, to pewnie twój ruch wytną te same mechanizmy, które wycinają boty teraz. Jeśli można byłoby robić tyle jednoczesnych requestów, co w api, to spoko :)

**Reply by adamz464** [reply]
ebay nie robi problemów bo każdy wie, do czego jest stosowany i że będzie to odświeżane automatami. Więc raczej nie oczekiwałbym tutaj problemów. Ew. sensowne ograniczenia typu aby nie odświeżać co 10s danego url to jest do akceptacji.

**Reply by maciekjaros** [reply]
Tak, tym bardziej że można zastosować cache requestów na wyższym poziomie i masowe odpytywanie nie będzie żadnym obciążeniem dla infrastruktury.

**Reply by adamz464** [reply]
no byle nie jak na OLX, że w zależności jak się trafiło - to były opóźnienia nawet do 20-30minut... bo leciało z cache :)

**Comment by meridPL** [comment]
A może drogie allegro, zrobisz jak część serwisów, które mają dane cenniejsze np. limit requestów na dzień ? Wydaje mi się, że rozwiązało by trochę czasu na scrapping, który generuje wam sztuczny ruch na stronie? Przecież jeżeli wchodzicie teraz na rynek europejski, to wypadałoby o tym pomyśleć :)

**Reply by hahyes** [reply]
Mało trafione z tym rynkiem europejskim, gdzie najwięksi gracze dają jeszcze bardziej upośledzony dostęp do danych niż Allegro np. Amazon

**Comment by adamz464** [comment]
jak się teraz skrapuje dane?
Selenium: "Confirm you are a human."
No nie mogę wybrnąć aby mnie przepuściło - czy to FF czy Chrome. Tryb rzeczywisty, żadne headless.

**Reply by meridPL** [reply]
Teraz to ciężko, allegro blokuje IP i teoretycznie co 15 sekund nie powinni Cię wychwycić. Większość VPN już jest poblokowanych.

**Reply by adamz464** [reply]
no nie do końca bo na swojej `normalnej` przeglądarce to allegro mogę nadal przeglądać (potwierdzę może z raz, że nie jestem botem i śmiga). W automacie to nie pomaga.

**Reply by DarkStoorM** [reply]
Jeżeli nie wysyłasz dużo zapytań, wystarczy w swoim curlu jako nagłówek wysłać to samo, co przeglądarka wysyła do Allegro i odpytać z tym wyszukiwarkę ;) nie potrzeba do tego żadnych tokenów, Allegro nie uzna Cię za bota, ale oczywiście, jak wyżej, w jakimś sensownym przedziale czasowym. Nie będę tego czasu testował, bo znając Allegro, zaraz zablokują permanentnie i będzie po pracy.

Nasze _drogie_ Allegro dopracowało swoją wyszukiwarkę w taki sposób, że nawet jak ręcznie porównuję produkty **w pracy** dzięki ich wyszukiwarce - **żadnych automatów**, wszystko na stronie Allegro - to mnie blokują, bo za dużo szukam :) jakaś paranoja normalnie.

Nie rozumiem. **Zamiast udostępnić nam okrojoną wersję tego samego zasobu**, to bawią się w jakieś głupoty, blokowania i weryfikacje. 
Czego nie wprowadzą, i tak to obejdziemy :)

**Reply by adamz464** [reply]
no to raczej do bani bo z curlem, przecież zaraz wyjdzie, że nie ma obsługi JS...

**Reply by DarkStoorM** [reply]
Ja akurat używam biblioteki HTTP z Laravel poprzez `Http::withHeaders(tutaj_nagłówek_przeglądarki)` w odpowiedzi dostaję prawidłowo pełny kod źródłowy z wynikiem wyszukiwania.

Niby to się niczym nie różni od wejścia na Allegro i skorzystania z wyszukiwarki, ale ich wyszukiwarka jest do bani i wolę sam sobie wyparsować oferty bez zbędnych pierdół i **poprawnym** sortowaniem. Nawet przy zakupach szukam wszystkiego _zewnętrznie_ :)

**Reply by adamz464** [reply]
dzięki, zbadam ten trop

**Reply by hahyes** [reply]
@DarkStoorM 

>Nasze drogie Allegro dopracowało swoją wyszukiwarkę w taki sposób, że nawet jak ręcznie porównuję produkty w pracy dzięki ich wyszukiwarce - żadnych automatów, wszystko na stronie Allegro - to mnie blokują, bo za dużo szukam :) jakaś paranoja normalnie.

Dziwisz się? Na portalu nie ma realnych zabezpieczeń w tym kierunku, a "sprawdzanie JS" sprowadza się do sprawdzenia nagłówków zapytań do serwera. Dobrze wiedzą, że scrapowanie tego portalu dla ogarniętego programisty to jest łatwizna, więc ograniczają ilościowo. Wychodzą z założenia, że zmiana wpływająca na garstkę osób zapewniająca niby bezpieczeństwo jest lepsza i szybsza (no i tańsza =). Ot i cała tajemnica. Allegro niestety w niektórych aspektach kładzie kompletnie lachę pod względem developmentu i wolą raczej zajmować się backendem. Frontend został pozostawiony marketingowcom i prawnikom, przez co nic nie ma tam sensu.

Chociaż muszę ich pochwalić - zrobili jakiś czas temu zasób we frontowym API do logowania się. Wcześniej to leciało po PHP i było upierdliwe (wchodzenie na stronę do pobrania odpowiednich ciastek, żeby wykonywać dalsze zapytania). Teraz to kwestia jednego zapytania i jesteś zalogowany oraz przygotowany w ciastka. Chociaż trochę ogarnęli ten burdel. :)

**Comment by janjablonsky** [comment]
> @adamz464  jak się teraz skrapuje dane?

bardzo łatwo. Można to zrobić na różne sposoby (nie - mnie nieblokują)
Na przykład mogę wykonać nawet tysiące requestow i bez zweryfikowanej aplikacji. I otrzymuję dane nawet gdybym miał dostęp do API (ten sam format)

> no to raczej do bani bo z curlem, przecież zaraz wyjdzie, że nie ma obsługi JS...

to nawet nie jest problem hhhh

> @DarkStoorM Nasze drogie Allegro dopracowało swoją wyszukiwarkę w taki sposób, że nawet jak ręcznie porównuję produkty w pracy dzięki ich wyszukiwarce - żadnych automatów, wszystko na stronie Allegro - to mnie blokują, bo za dużo szukam :) jakaś paranoja normalnie.

Tak "żadnych automatów,", mam automat i mogę wygenerować tyle tokenów, ile chcę ;) - w każdej chwili
Demo:
<a href="https://ibb.co/BryxX8h"><img src="https://i.ibb.co/BryxX8h/demo2.gif" alt="demo2" border="0"></a>

**Comment by janjablonsky** [comment]
@juhacz, @adamz464 jak Allegro nie bedzie robic stresu to moge opublikowac kod od automatu (do generowania access_token) + 1 metode do scrapowania (opis jak to zrobic + PHP przyklad), ktora dziala i moge przeczytac:
- /offers/listing informacje (wyszukiwarka -> wkladasz phrase ...)
- informacje o dostawie produktu (kurier na adres, ...)
...
wszystko JSON format, kompatybilny z Allegro API

automat do generowania access_token umie:
- latwo pobija allegro weryfikacja 
- generuje access_token
- zobacz demo https://github.com/allegro/allegro-api/discussions/4584#discussioncomment-1896239
...
tak samo, wszystko JSON format, kompatybilny z Allegro API

**Reply by meridPL** [reply]
Hej, 
Jak nie opublikujesz tutaj, to odezwij się na kontakt@qu2qu.ct8.pl :)

**Reply by Kpichurski** [reply]
do mnie też się możesz odezwać email: crisof9@gmail.com

**Reply by meridPL** [reply]
Do mnie się nie odezwał więc już zwątpiłem :)

**Reply by hahyes** [reply]
Podejrzewam, że albo to załatali, albo chciał oscamować ludzi tutaj. Wątpię, aby była metoda na "generowanie" access tokenów. W sumie brzmi to nierealnie, żeby być szczerym.

**Reply by janjablonsky** [reply]
@hahyes  "...albo chciał oscamować ludzi tutaj." dziekuje za opinie (P.S. demo.gif bylo tylko 3 tygodnie dostepne). Doceniam, ale nie obchodza mnie $.
Nie mialem czas zająć się tą sprawą, jest to tylko dla mnie "side-project" i w sumie interesuje mnie zabezpieczenie systemu.

W tym przypadku zostawimy ta sprawe przy scrapowaniu, to nikomu nieskodzi (nawet allegro niebedzie robic stresu ;) )

Przykladowa metoda scrapowania:
Nawet allegro za ciebie rozparsuje wszystkie dane z wyszukiwarki (https://allegro.pl/listing?string=<phrase\>).
Czyli bierzesz strone https://allegro.pl/listing?string=<phrase\>. (sortowanie &order=p i tak dalej)
- Szukas //div[data-prototype-id="allegro.listing"] no i bierzesz wartosc atributu data-box-id 
- dalej szukas tylko //script[data-serialize-box-id=<wartosc atributu data-box-id\>], no i to jest JSON z danymi (__listing_StoreState)

Informacje o produkcie - Bierzesz strone produktu
- Szukas //div[data-box-name="ShippingInfoShowSpinner" and data-placeholder-for i bierzesz wartosc atributu data-placeholder-for 
- dalej szukas tylko //script[data-serialize-box-id=<wartosc atributu data-placeholder-for \>], no i tam jest JSON z danymi tego produktu (opboxContext)
- Informacje o dostawie produktu: powtarzasz ten sam request tylko zamienasz naglowek accept za "accept: application/vnd.opbox-web.subtree+json" i wkladasz nowy "x-box-id: \<wartosc atributu data-placeholder-for >"
- odzyskujesz JSON

Milego dnia i zabawy

**Reply by hahyes** [reply]
> @hahyes "...albo chciał oscamować ludzi tutaj." dziekuje za opinie (P.S. demo.gif bylo tylko 3 tygodnie dostepne). Doceniam, ale nie obchodza mnie $. Nie mialem czas zająć się tą sprawą, jest to tylko dla mnie "side-project" i w sumie interesuje mnie zabezpieczenie systemu.
> 
> W tym przypadku zostawimy ta sprawe przy scrapowaniu, to nikomu nieskodzi (nawet allegro niebedzie robic stresu ;) )
> 
> Przykladowa metoda scrapowania: Nawet allegro za ciebie rozparsuje wszystkie dane z wyszukiwarki (https://allegro.pl/listing?string=<phrase>). Czyli bierzesz strone https://allegro.pl/listing?string=<phrase>. (sortowanie &order=p i tak dalej)
> 
> * Szukas //div[data-prototype-id="allegro.listing"] no i bierzesz wartosc atributu data-box-id
> * dalej szukas tylko //script[data-serialize-box-id=<wartosc atributu data-box-id>], no i to jest JSON z danymi (__listing_StoreState)
> 
> Informacje o produkcie - Bierzesz strone produktu
> 
> * Szukas //div[data-box-name="ShippingInfoShowSpinner" and data-placeholder-for i bierzesz wartosc atributu data-placeholder-for
> * dalej szukas tylko //script[data-serialize-box-id=<wartosc atributu data-placeholder-for >], no i tam jest JSON z danymi tego produktu (opboxContext)
> * Informacje o dostawie produktu: powtarzasz ten sam request tylko zamienasz naglowek accept za "accept: application/vnd.opbox-web.subtree+json" i wkladasz nowy "x-box-id: <wartosc atributu data-placeholder-for >"
> * odzyskujesz JSON
> 
> Milego dnia i zabawy

demo.gif umarło po dwóch tygodniach jak nie szybciej. Spójrz, kiedy odpisywałem, wtedy gif był martwy.
A zescrapować z wyszukiwarki na stronie to każdy średnio rozgarnięty zrobi. Jednak jest różnica między scrapowaniem publicznej wyszukiwarki na stronie a pisaniem dziwnych tekstów bez pokrycia o generowaniu tokenów i korzystaniu z /offers/listing. Chyba, że ci coś się pomyliło i ci nie chodzi o API, tylko o sam portal. To wtedy nie wiem, po co ci jakieś tokeny generować.

Bardzo mi przykro, że nie daje wiary magicznym sposobom, które brzmią jak scam z daleka.

**Comment by Michalzip** [comment]
Czy wątek doszedł do zamierzonego celu? bo próbuje uzyć tego endpointu i widze że nic w tej sprawie nie zrobili...

**Comment by juhacz** [comment]
Prędzej nowy premier zrealizuje to co obiecał, niż Allegro udostępnij ponownie ten endpoint dla wszystkich.

**Comment by JedrzejMakuch** [comment]
Podbijam znow, Allegro po co wam te sugestie, skoro nawet nie odpowiadacie?

**Comment by MartaNowaczyk** [comment]
Sugestie zbieramy i na bieżąco przekazujemy do odpowiednich zespołów. Gdy coś dopisujecie dodatkowo monitorujemy sprawę i potwierdzamy poprawność statusu. Na ten moment nie podjęto decyzji o tym by wdrożyć proponowaną zmianę, stąd jest w fazie oczekiwania. Jeśli jednoznacznie będzie wskazanie, że pomysł jest przyjęty lub odrzucony na pewno zmienimy odpowiedni status i damy znać.

**Comment by MartaNowaczyk** [comment]
Zaktualizowaliśmy status sugestii jako "odrzucona", ponieważ aktualnie zasobu GET /offers/listing nie ma w publicznym API.
